models/dataset.go

```diff
@@ -457,21 +457,54 @@ func ValidateDataset(ctx context.Context, dataset *Dataset) error {
 	var invalidFields []string
 	if dataset.URI != "" {
 		dataset.URI = strings.TrimSpace(dataset.URI)
-		datasetURI, err := url.Parse(dataset.URI)
+		_, err := url.Parse(dataset.URI)
 		if err != nil {
 			invalidFields = append(invalidFields, "URI")
 			log.Event(ctx, "error parsing URI", log.ERROR, log.Error(err))
-		} else {
-			if !strings.Contains(datasetURI.Path, "/datasets") || !strings.Contains(datasetURI.Path, dataset.ID) {
-				fmt.Println("hello")
-				invalidFields = append(invalidFields, "URI")
+		}
+	}
+
+	for i := range dataset.Publications {
+		generalDetails := &dataset.Publications[i]
+		if generalDetails != nil {
+			generalDetails.HRef = strings.TrimSpace(generalDetails.HRef)
+			_, err := url.Parse(generalDetails.HRef)
+			if err != nil {
+				invalidFields = append(invalidFields, "href")
+				log.Event(ctx, "error parsing URI", log.ERROR, log.Error(err))
+			}
+		}
+	}
```

Comment:
This logic is duplicated at least 3 times which to me suggests that we have common functionality which we could abstract into a more generic form to allow reuse and reduce duplication.

From what I can see the only difference in each case is the `type` we are operating on - I think. Please point out if I have missed something. If this is the case I think we could achieve a more extendable solution if we use an interface.  I'm thinking something like:

```
type interface Trimmable {
    GetHref() string
    SetHref(val string)
}
```
Then we can implement these 2 methods on each struct type we want to trim. Taking `Publications` as an example add the following:
```
func (p Publication) GetHref() string {
    return p.HRef
}
....
func (p *Publication) SetHref(s string) {
    p.HRef = s
}
```
We can then create 1 function for trimming whitespace that takes the `Trimmable` type .
```
func (input []Trimmable) {
	for _, item := range input{ 
         //  GetHref() to get the value, SetHref() to update with a parsed value
	}
}
```
I think something like this should work. If we need to `trim` more types in future we can simple implement the interface on the type and it can be passed into this existing function.